File: src/World.ts
import Biome from "./Biome.js";
import Chunk from "./Chunk.js";
import Grid from "./Grid.js";
import Perlin from "./Perlin.js";
import Rangerray from "./Rangerray.js";
import { BASE_BIOME_SIZE, CHUNK_SIZE, GRID_IMAGE_FILENAMES, RENDERER } from "./constants.js";
import { leftJustify, objectFromEntries } from "./functions.js";
import { existsSync, join, loadJSON, mkdirSync, writeFileSync } from "./env_script.js";
import { Colour, GridImageName, WarningRecord, WorldConfig, WorldInfo } from "./types.js";

class World {

    public filepath: string;
    public name: string;
    public config: WorldConfig;
    public renderWorld: boolean = true;
    public widthInTiles: number;
    public heightInTiles: number;
    public widthInChunks: number;
    public heightInChunks: number;
    public seed: string;
    public minNoise: number;
    public maxNoise: number;
    public totalHeight: number;
    public maxHeight: number;
    public totalAreaInTiles: number;
    public tempAcc: number;
    public tempCount: number;
    public noiseAcc: number;
    public noiseCount: number;
    public warningRecord: WarningRecord;
    public biomesRangerray: Rangerray<Biome>;
    public biomeSize: number = 1;
    public biomeSuperGridTileSize: number = 1;
    public biomeSizes!: { [index: string]: number };
    public biomeNames: string[] = [];
    public biomeColours: { [index: string]: string };
    public worldInfo: WorldInfo;

    public constructor(name: string, filepath: string) {

        this.filepath = filepath;
        this.name = name;
        let configFilePath = [this.filepath, "CONFIG.json"].join("/");
        this.config = loadJSON<WorldConfig>(configFilePath);

        this.fillConfig();

        this.seed = this.config.seed.toString();
        this.widthInChunks = this.config.width;
        this.heightInChunks = this.config.height;
        this.widthInTiles = this.widthInChunks * CHUNK_SIZE;
        this.heightInTiles = this.heightInChunks * CHUNK_SIZE;
        this.totalAreaInTiles = this.widthInTiles * this.heightInTiles;
        this.createSaveFiles();

        this.warningRecord = { maxHeight: [], matchingBiomeColours: [] };
        this.biomeColours = {};

        this.maxHeight = Math.floor(this.config.maxHeight);
        this.totalHeight = 2 * this.maxHeight
        this.tempAcc = 0;
        this.tempCount = 0;

        this.biomesRangerray = new Rangerray("biomesRangerray");

        this.configureBiomes()

        // stats
        this.minNoise = 1;
        this.maxNoise = 0;
        this.noiseAcc = 0;
        this.noiseCount = 0;

        this.worldInfo = {
            seed: this.seed,
            width: this.widthInChunks,
            height: this.heightInChunks,
            q: this.config.q,
            r: this.config.r,
            maxHeight: this.maxHeight,
            totalHeight: this.totalHeight
        };

        this.generateChunks();

        writeFileSync([this.filepath, "GENERATED", "WORLD_INFO.json"].join("/"), JSON.stringify(this.worldInfo));

    }

    private fillConfig(): void {
        if (this.config.q === undefined) this.config.q = 0;
        if (this.config.r === undefined) this.config.r = 0;
    }


    public createSaveFiles() {
        if (!existsSync(this.filepath)) {
            mkdirSync(this.filepath);
            mkdirSync([this.filepath, "GENERATED", "images"].join("/"));
            mkdirSync([this.filepath, "GENERATED", "chunks"].join("/"));
        }
    }

    public configureBiomes() {

        this.biomeNames = this.config["biomes"].map((biome: [number, string]) => biome[1]);
        this.biomeSizes = objectFromEntries<number>(this.biomeNames, new Array(this.biomeNames.length).fill(0));

        let keys = Object.keys(this.config);

        if (keys.includes("biomeSize")) {
            this.biomeSize = this.config["biomeSize"];
        }
        else {
            this.biomeSize = 1;
        }

        let biomeSizeSum = this.config.biomes.reduce((acc: number, x: any) => acc + x[0], 0);

        Rangerray.fracrrayToRangerray(this.config["biomes"]);

        for (let biomeEntry of this.config["biomes"]) {
            let upperPoint = biomeEntry[0];
            let biomeName = biomeEntry[1];
            upperPoint = Perlin.flatten(upperPoint);
            let biome = new Biome(biomeName, this);
            this.biomesRangerray.insert(upperPoint, biome);
        }

        this.biomeSuperGridTileSize = biomeSizeSum * BASE_BIOME_SIZE * this.biomeSize;

        console.log(`Biome super grid tile size: ${this.biomeSuperGridTileSize}`);

    }

    public generateChunks() {

        let gridImageNames: GridImageName[] = (this.renderWorld) ? ["surfaceGridImage", "biomeGridImage", "subBiomeGridImage", "perlinImage"] : [];

        let gridImages: { [index: string]: Grid<Colour> } = {};
        for (let gridImageName of gridImageNames) {
            gridImages[gridImageName] = new Grid<Colour>(this.widthInTiles, this.heightInTiles, [0, 0, 0]);
        }

        for (let _q = 0; _q < this.widthInChunks; _q++) {
            for (let _r = 0; _r < this.heightInChunks; _r++) {
                let chunk = new Chunk(this, this.config.q + _q, this.config.r + _r);
                let cornerX = _q * CHUNK_SIZE;
                let cornerY = _r * CHUNK_SIZE;
                for (let gridImageName of gridImageNames) {
                    gridImages[gridImageName].overlay(chunk[gridImageName], cornerX, cornerY);
                }
            }
        }

        for (let gridImageName of gridImageNames) {
            let filename = GRID_IMAGE_FILENAMES[gridImageName];
            RENDERER.renderColourGrid(gridImages[gridImageName], join(this.filepath, "GENERATED", "images", filename));
        }

    }

    public summarise() {
        console.log(`Biome average: ${this.tempAcc / this.tempCount}`);
        for (let biomeName of Object.keys(this.biomeSizes)) {
            let biomeSize = this.biomeSizes[biomeName];
            let percentage = 100 * biomeSize / this.totalAreaInTiles;
            console.log(`${leftJustify(biomeName, 20)} ${leftJustify(biomeSize.toString(), 10)} ${percentage}%`);
        }
    }

}

export default World;

File: src/World.py
import json, math, os, sys

from src.constants import BASE_BIOME_SIZE, CHUNK_SIZE

from src.functions import exit_with_error, flatten_noise_distribution, point_at_portion_between, raise_warning, save_json

from src.SubBiome import SubBiome
from src.Grid import Grid
from src.Rangerray import Rangerray
from src.Chunk import Chunk

class World:

    def __init__(self, name, config, render_world):

        self.name = name
        self.config = config
        self.render_world = render_world
        self.seed = config["seed"]
        self.width_in_chunks, self.height_in_chunks = config["width"], config["height"]
        self.width_in_tiles = self.width_in_chunks * CHUNK_SIZE
        self.height_in_tiles = self.height_in_chunks * CHUNK_SIZE
        self.total_area_in_tiles = self.width_in_tiles * self.height_in_tiles
        self.create_save_files()

        self.warnings_raised = {
            "maxHeight": [],
            "matching_biome_colours": []
        }
        self.biome_colours = dict()

        self.maxHeight = int(self.config["maxHeight"])
        self.total_height = 2 * self.maxHeight
        self.temp_acc, self.temp_count = 0, 0

        self.configure_biomes()

        # stats
        self.min_noise, self.max_noise = 1, 0
        self.noise_acc, self.noise_count = 0, 0

        self.world_info = {
            "seed": self.seed,
            "width": self.width_in_chunks,
            "height": self.height_in_chunks,
            "maxHeight": self.maxHeight,
            "total_height": self.total_height
        }

        self.generate_chunks()

        save_json(self.world_info, os.path.join("worlds", self.name, "WORLD_INFO.json"))

        self.summarise()



    def create_save_files(self):
        filepath = os.path.join("worlds", self.name)
        if not os.path.exists(filepath):
            os.makedirs(filepath)
            os.makedirs(os.path.join(filepath, "images"))
            os.makedirs(os.path.join(filepath, "chunks"))

    def configure_biomes(self):
        self.biomes_rangerray = Rangerray("biomes_rangerray")
        lower_point = 0
        self.biome_names = [biome[1] for biome in self.config["biomes"]]
        self.biome_sizes = dict(zip(self.biome_names, [0]*len(self.biome_names)))

        if "biome_size" in self.config.keys():
            self.biome_size = self.config["biome_size"]
        else:
            self.biome_size = 1

        Rangerray.fracrray_to_rangerray(self.config["biomes"])

        for biome in self.config["biomes"]:
            upper_point, biome_name = biome[0], biome[1]
            upper_point = flatten_noise_distribution(upper_point)
            rangerray = self.create_biome(biome_name, lower_point, upper_point)
            self.biomes_rangerray.insert(upper_point, rangerray)
            lower_point = upper_point

        self.biome_super_map_tile_size = len(self.config["biomes"]) * BASE_BIOME_SIZE * self.biome_size


    def create_biome(self, biome_name, biome_noise_lower, biome_noise_upper):
        rangerray = Rangerray(biome_name)
        biome_config_path = os.path.join("configs", self.name, "biomes", biome_name + ".json")
        file = open(biome_config_path, "r")
        biome_config = json.load(file)
        noise_lower, noise_upper = 0, 0

        if tuple(biome_config["colour"]) in self.biome_colours.values():
            raise_warning("Matching biome colours", "The biome " + biome_name + " is using a colour already in use.")
        self.biome_colours[biome_name] = tuple(biome_config["colour"])

        Rangerray.fracrray_to_rangerray(biome_config["ranges"])

        for sub_biome in biome_config["ranges"]:
            noise_upper, sub_biome_name = sub_biome[0], sub_biome[1]
            if sub_biome_name not in biome_config:
                exit_with_error("Undefined sub-biome", "An undefined sub-biome named " + sub_biome_name + " is referenced inside ranges attribute of biome " + biome_name + ".")
            noise_upper = flatten_noise_distribution(noise_upper)
            obj = SubBiome(self, biome_name, sub_biome_name, biome_config, noise_lower, noise_upper)
            rangerray.insert(noise_upper, obj)
            noise_lower = noise_upper

        return rangerray


    def generate_chunks(self):

        if self.render_world:
            map_image_names = ["surface_map_image", "biome_map_image", "sub_biome_map_image", "perlin_image"]
        else:
            map_image_names = []
        grids = [Grid(self.width_in_tiles, self.height_in_tiles, 0) for _ in map_image_names]

        map_images = dict(zip(map_image_names, grids))

        for q in range(self.width_in_chunks):
            for r in range(self.height_in_chunks):
                chunk = Chunk(self, q, r)
                corner_x, corner_y = q * CHUNK_SIZE, r * CHUNK_SIZE
                for map_image_name in map_image_names:
                    map_images[map_image_name].overlay(getattr(chunk, map_image_name), corner_x, corner_y)

        for map_image_name in map_image_names:
            map_images[map_image_name].save_RGBs(self.name + "_" + map_image_name, self.name)

    def summarise(self):
        print("Biome average = " + str(self.temp_acc / self.temp_count))
        for biome_name, biome_size in zip(self.biome_sizes.keys(), self.biome_sizes.values()):
            percentage = 100 * biome_size / self.total_area_in_tiles
            print(str.ljust(biome_name, 20) + "\t" + str.ljust(str(biome_size), 10), str(percentage) + "%")

